File: data_structures/py/hackerrank_2dArray.py
# HACKER RANK - https://www.hackerrank.com/challenges/2d-array/problem


#!/bin/python3

import math
import os
import random
import re
import sys

def getTopWindow(a, i, j):
    return a[i][j] + a[i][j+1] + a[i][j+2]

def getMidWindow(a, i, j):
    return a[i+1][j+1]

def getBottomWindow(a, i, j):
    return getTopWindow(a, i+2,j)

# Complete the hourglassSum function below.
def hourglassSum(arr):
    
    sums = []
    for i in range(0,4):
        for j in range(0,4):
            sums.append(getTopWindow(arr, i, j) + getMidWindow(arr, i, j) + getBottomWindow(arr, i, j))
            
    return max(sums)

if __name__ == '__main__':
    fptr = open(os.environ['OUTPUT_PATH'], 'w')

    arr = []

    for _ in range(6):
        arr.append(list(map(int, input().rstrip().split())))

    result = hourglassSum(arr)

    fptr.write(str(result) + '\n')

    fptr.close()

File: data_structures/cpp/hackerrank_2dArray.cpp
// HACKER RANK - https://www.hackerrank.com/challenges/2d-array/problem

#include <bits/stdc++.h>

using namespace std;
int getTopWindow(vector<vector<int>> & arr, const size_t i, const size_t j) {
    return arr[i][j] + arr[i][j+1] + arr[i][j+2] ;
}

int getMidWindow(vector<vector<int>> & arr, const size_t i, const size_t j) {
    return arr[i+1][j+1]  ;
}

int getBottomWindow(vector<vector<int>> & arr, const size_t i, const size_t j) {
    return getTopWindow(arr, i+2, j) ;
}
// Complete the hourglassSum function below.
int hourglassSum(vector<vector<int>> arr) {

    vector<int> sums;
    for (size_t i = 0; i < 4; ++i) {
        for (size_t j = 0; j < 4; ++j) {
            sums.push_back(getTopWindow(arr, i, j) + 
                          getMidWindow(arr, i, j) + 
                          getBottomWindow(arr, i, j));
        }
    }
    
    // find max
    int max = INT_MIN;
    for (size_t it = 0; it != sums.size(); ++it) {
        if (sums[it] > max)
            max = sums[it];
    }
    
    return max;

}

int main()
{
    ofstream fout(getenv("OUTPUT_PATH"));

    vector<vector<int>> arr(6);
    for (int i = 0; i < 6; i++) {
        arr[i].resize(6);

        for (int j = 0; j < 6; j++) {
            cin >> arr[i][j];
        }

        cin.ignore(numeric_limits<streamsize>::max(), '\n');
    }

    int result = hourglassSum(arr);

    fout << result << "\n";

    fout.close();

    return 0;
}

